RFC: Allow packages to specify a set of supported targets (#3759)
rust-lang/rfcs: text/3759-cargo-supported-targets.md

Reviewer: @workingjubilee at https://github.com/rust-lang/rfcs/pull/3759#issuecomment-2581161497

> Maybe if we adopt this it should include a way to describe whether the use of the crate on an unsupported target is one of
> 
>     * actively dangerous (implementations depend on target-specific details that _must_ be ported)
> 
>     * probably bad? but not _automatically_ unsound (implementation depends on target-specific details that work if another target is _similar enough_)
> 
>     * maybe it works, maybe it doesn't? glwt (the common "I've only tested it on..." situation)

*(moved here to move this off the main thread so the discussion can be followed)*

Reviewer: Looking at the Motivation section (and from my understanding of the problem), this is meant to help with
- `cargo check --workspace` when some packages aren't supported
- Smaller `Cargo.lock` / `vendor/` 

In those situations, the main reason for a package to limit the targets is
- a workspace member can't build on all targets
- An application will only ever be built for specific targets and they want dependencies stripped that aren't for that target
- A library is platform-specific (e.g. `windows-sys`) and they want to allow themselves to be stripped for the above.

imo this shouldn't be about "I only tested on this platform".  I'm not even sure if we should get nuanced about the other two cases.  I also don't think distinguishing these cases is worth the data format and end-user complexity.

I do think we should be clear in communicating abut this how this is intended to be used to reduce problems with it.  For example, the name alone invites a lot of this kind of thinking which I noted at https://github.com/rust-lang/rfcs/pull/3759#discussion_r1909408062

Reviewer 2: Perhaps it should be clearer that it's `allowed-targets`, i.e. that all others are implicitly *banned*.

Author: Maybe `required-targets` makes it a bit more clear?

Reviewer 3: a more obtuse option that might make desired usage clearer would be to have `known-broken-targets` and require specifying where it doesn't work, instead of where it works.

Reviewer: Might be good to consolidate naming conversations to https://github.com/rust-lang/rfcs/pull/3759#discussion_r1909195446

Reviewer: This is similar to the "support" discussion at https://github.com/rust-lang/rfcs/pull/3759#discussion_r1909379623

> The RFC scope has been narrowed down to just package/build-target filtering when running Cargo commands and the field is stripped on publish to prevent it from being misused. Erroring has been moved to a future possibility and we can discuss it further when/if someone proposes that in the future.

Reviewer: Is there a feature needed for proc-macros in a workspace like this?  With #14903, this would also extend to workspace members that exist to be build scripts.

I'm wondering if there is any design overlap here that could benefit from considering proc-macros / build script `cargo check --workspace` filtering in the future.

Author: Yes I think this is a viable option. I'm not sure how this should be specified but it is clearly relevant.

Author: This comment is not directly related but I think that a proc-macro is like a build script in that it _theoretically_ should allow any target to build it, because it might be run on any host imaginable.

Reviewer: Yes, a proc-macro is buildable in any host-supported target.

What I'm trying to highlight is if I run `cargo check --target <tuple> --workspace`, then a proc-macro will be built twice
- Once for the host toolchain for being used in the workspace members that depend on it
- Once for the target toolchain because that was requested

I was wondering if there could be a way for people to opt-in to `cargo check --target <tuple> --workspace` only building the proc-macro once.  This RFC has some similar concepts (skipping of packages based on platform-related info) which is why I was wondering if (1) people find value in this idea and (2) if its close enough to this that we should see if we can design with this in mind as a future possibility

- Feature Name: `supported-targets`
- Start Date: 2025-01-08
- Pre-RFC: [Rust
  internals](https://internals.rust-lang.org/t/pre-rfc-allow-packages-to-specify-a-set-of-supported-targets/21979)
- RFC PR: [rust-lang/rfcs#3759](https://github.com/rust-lang/rfcs/pull/3759)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

The word _target_ is extensively used in this document. The
[glossary](https://doc.rust-lang.org/cargo/appendix/glossary.html#target) defines its many meanings.
Here, _target_ refers to the "Target Architecture" for which a package is built. Otherwise, the
terms "cargo-target" and "target-triple" are used in accordance with their definitions in the
glossary.

> # Summary
> 
> The addition of `supported-targets` to `Cargo.toml`. This field is a `cfg` string that restricts the
> set of targets which a package supports. Packages can only be built for targets that satisfy their
> `supported-targets`.
> 

Reviewer: Its helpful to have an example in the Summary
````suggestion
# Summary

The addition of `supported-targets` to `Cargo.toml`. This field is a `cfg` string that restricts the
set of targets which a package supports. Packages can only be built for targets that satisfy their
`supported-targets`.

```toml
[package]
name = "hello_cargo"
supported-targets = 'cfg(any(target_os = "linux", target_os = "macos"))'
```

````

> # Motivation
> 
> Some packages rely on features or behavior that is not available on every platform `rustc` can target. Currently, there is no way to formally

Reviewer: It would be good to link to the original issue somewhere, e.g.

```suggestion
# Motivation

_For more background, see [rust-lang/cargo#6179](https://github.com/rust-lang/cargo/issues/6179)_

Some packages rely on features or behavior that is not available on every platform `rustc` can target. Currently, there is no way to formally
```

When a `cargo` build command (e.g. `check`, `build`, `run`, `clippy`) is run, it checks that the
selected target satisfies the `supported-targets` of the package being built. If it does not, the
package is skipped or an error is raised, depending on how [`cargo` was invoked](ignoring-builds).
However, `supported-targets` is _only_ checked for commands that take a `--target` option and does not affect other commands (e.g., `cargo fmt`).

As this field is limited to local development, `cargo package` / `cargo publish` will strip it from `Cargo.toml`.
Including the field in the `.crate` file is left as a [future possibility](#future-possibilities) for now.

## Ignoring builds for unsupported targets
[ignoring-builds]: #igonring-builds-for-unsupported-targets

If cargo is invoked in a workspace or virtual workspace without specifying a package as
build-target, then `cargo` skips any package that does not support the selected target. If a package
is specified using `--package` or if `cargo` is invoked on a single package, and the selected target
does not satisfy the `supported-targets` of the package, then an error is raised.

> # Drawbacks

Reviewer: @nsabovic  at https://github.com/rust-lang/rfcs/pull/3759#issuecomment-2596936489

> However, I think that this is OK because it buys us capability-based checks vs. version-based checks. This is a big deal for me. For a flowery exposition, read Linus Torvald's posts about x86 feature levels. In our own code, the lack of platform checks has been a forcing function which caused us to have separate features for e.g. `zbus`, `rpi-gpio` and `termios2` whereas if we had the hammer of `cfg(target_os = "linux")` we probably would have used that and gotten brittle, non-future-proof code. For example, `zbus` works on macOS and if we needed to use it for some tests and didn't control the crate, with the proposed solution we would have probably been out of luck. So in a way I feel this has a potential to subvert cargo features and this is something we can't fix with a flag.

Reviewer: @ssokolow at https://github.com/rust-lang/rfcs/pull/3759#issuecomment-2597921080

> > However, I think that this is OK because it buys us capability-based checks vs. version-based checks. This is a big deal for me. For a flowery exposition, read Linus Torvald's posts about x86 feature levels.
> 
> Or the entire history of the web ecosystem for that matter. There was a _lot_ of effort poured into getting people to do JavaScript feature checks instead of stuff like User-Agent string checks.
> 
> Hell, that's why Mobile Chrome's User-Agent string is `Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Mobile Safari/537.36`.
> 
> It's literally Chrome pretending to be Mobile Safari in the name of compatibility, which is careful to mention that it's WebKit in the name of compatibility, which is pretending to still be its ancestor KHTML in the name of compatibility, which is pretending to be Gecko in the name of compatibility, which is careful to still announce itself as Mozilla (i.e. Netscape Navigator. NCSA Mosaic-zilla) in the name of compatibility, just as Internet Explorer did right from version 1.0 as far as I can tell without spinning up a copy in an emulator to confirm. One big accretion over time. The [most trustworthy UA list I found](https://www.zytrax.com/tech/web/msie-history.html) only goes back to IE 2.0 claiming to be Mozilla/1.22. (Yes, the reason [_every_](https://www.useragents.me/) desktop and mobile browser User-Agent string begins with `Mozilla` predates the "Best Viewed with Netscape Navigator" badges of the 90s.)

Reviewer: Like the conversations around "support" (e.g. https://github.com/rust-lang/rfcs/pull/3759#discussion_r1909403653), let's keep in mind that the primary focus of this is for affecting the developer who sets this field.  The RFC has been scoped down to build-target filtering with `cargo check --workspace` and the primary goals after that are `cargo vendor` / `Cargo.lock` pruning.  We also have the possibility of improving error quality when depending on a package in a way that doesn't meet its requirements.  iiuc this conversation mostly focuses on that last point which is no longer in this RFC and what matters right now is whether we can support this feature or not in the future.  Whether we should or not, and how, is for a future discussion.  I think this kind of roll out will also help set the tone for this feature.  iirc the RFC also dropped specifying specific targets.  Within `cfg`, people can focus in on the specific requirements they have (e.g. pointer width) or can use `features`, depending on what best fits their purpose.

[drawbacks]: #drawbacks

- Once implemented this does not yet solve the problem of dependency pruning in `Cargo.lock`, which
  is a common use case for this feature.
- This is the first step towards a target aware `cargo`, which may increase `cargo`'s complexity,
  and bring more feature requests along these lines.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

> ## Format

Reviewer: I thought this came up in one of the Issues but I'm not finding it atm.

That proposed schema is specifying conditional membership in the workspace which is not a mental model that aligns well with Cargo imo.  Workspace members are always members.  Build targets are always build targets.  Whats changing is the conditions under which they are processed by default.

This is also awkward to work with as data about a package is being stored at the workspace level.

This would preclude the error reporting future possibility because we are not longer describing the requirements of a package but the conditions under which it is a member.

See also: [using a package vs. using a workspace](package-vs-workspace).

[package-vs-workspace]:
#https://blog.rust-lang.org/inside-rust/2024/02/13/this-development-cycle-in-cargo-1-77.html#when-to-use-packages-or-workspaces

## Not using `cfg`

Using `cfg` complicates the implementation (and thus maintenance), and may require a substantial
amount of calls to `rustc` to check target-`cfg` compatibility. Some alternatives are discussed here
along with their drawbacks.

### Using wildcards

Instead of using `cfg` specifications, one could use wildcards (e.g., `x86_64-*-linux-*`). This is
much simpler to implement, target-triples are syntactically checked for a match instead of solving
set relations for `cfg`. However, this is not as expressive as `cfg`, and does not correctly
represent the semantics of target-triples. For example, supporting `target_family = "unix"` would
require an annoyingly long list of wildcard patterns. Things like `target_pointer_width = "32"` are
even harder to represent, and things like `target_feature = "avx"` are basically not representable.
Also, this is new syntax not currently used by cargo.

> ### Allowing only target-triples

Reviewer: As target-tuples were removed, we should also have section talking about including them

would leave it up to the user to infer it. Some crates also made use of compile time errors to
ensure that `cfg` requirements are met, for example:
```rust
#[cfg(not(any(…)))]
compile_error!("unsupported target cfg");
```
[`getrandom`](https://github.com/rust-random/getrandom/blob/9fb4a9a2481018e4ab58d597ecd167a609033149/src/backends.rs#L156-L160)
is an example of a crate utilizing this method.

In other system level languages, vendoring dependencies is a common practice, and the user would be
responsible for ensuring that the dependencies are compatible with the target. For interpreted
languages, this is a non-issue because any platform being able to run the interpreter can run the
package.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- Should we strip the `cfg` prefix from the field e.g., `supported-targets = 'target_os = "linux"'`?

# Future possibilities
[future-possibilities]: #future-possibilities

## Ensuring proper use of dependencies

By leveraging compilation errors, it would be possible to ensure that a package only uses
dependencies that support the package's `supported-targets`. This would be done by comparing the

> `supported-targets` of the package with the `supported-targets` of the dependencies.

Reviewer: imo, like the motivation section, this should focus on improving error messages and detection incompatible dependencies without exhaustively testing on every platform

> `supported-targets` of the package with the `supported-targets` of the dependencies.

Reviewer: imo, like the motivation section, this should focus on improving error messages and detection incompatible dependencies without exhaustively testing on every platform

Author: Fixed in https://github.com/rust-lang/rfcs/pull/3759/commits/0b21a0aeae4c633d9b572d709bee418a720c61fe.

Cargo's documentation should give clear guidance for when to use this field, and should not suggest
using it by default. In particular, we should steer users to use this when they have good reason to
believe the crate will not compile or work as expected (e.g. because it uses target-specific APIs),
and not use it merely for "I haven't personally tested this on other targets".

> ### Compatibility of `[dependencies]`

Reviewer: Ideally, we don't lose information from this discussion.  In this case, we should include the risk that `package.supported-targets` is too restrictive and include a summary of some of the ways of resolving it and how some may interact poorly with `Cargo.lock` trimming.

See https://github.com/rust-lang/rfcs/pull/3759#discussion_r1910655388, https://github.com/rust-lang/rfcs/pull/3759#discussion_r1909402886

Author: Added in https://github.com/rust-lang/rfcs/pull/3759/commits/b64731dc698958a9078c53bbaa5ced17c7fd20ab.

Reviewer: That didn't include all of the options discussed

Author: I added a [paragraph](https://github.com/rust-lang/rfcs/pull/3759/files#diff-7ca74978b782c1ea19b9cafec0cd71f3936a4791ac30656f361896ee217d391cR298-R303) that discusses the interoperability of lockfile trimming & ignoring errors/lints, was there anything else I missed?

Reviewer: Closing in favor of https://github.com/rust-lang/rfcs/pull/3759/files#r1925403294

dependency does not have a `target` field, then it would be checked against the package's
`supported-targets`, like any other dependency.

## Eliminating unused dependencies from `Cargo.lock`

A package's dependencies may themselves have `[target.'cfg(..)'.dependencies]` tables, which may
never be used because of the `supported-targets` restrictions of the package. These can safely be
eliminated from the dependency tree of the package.

Consider the following example:
```toml
[package]
name = "foo"
# ...
supported-targets = 'cfg(target_os = "linux")'

[dependencies]
bar = "0.1.0"
```
```toml
[package]
name = "bar"

[target.'cfg(target_os = "macos")'.dependencies]
baz = "0.1.0"
```
Currently, `baz` is included in the dependency tree of `foo`, even though `foo` is never built for
`macos`. `baz` could be pruned from the dependency tree of `foo`, since `target_os = "macos"` is
mutually exclusive with `target_os = "linux"`.

Formally, dependencies (and transitive dependencies) under `[target.**.dependencies]` tables are
eliminated from the dependency tree of a package if the `supported-targets` of the package is
mutually exclusive with the target preconditions of the dependency.

> ## Comparing `supported-targets`

Author: This was a mistake, fixed in https://github.com/rust-lang/rfcs/pull/3759/commits/ef86a06d8be5e35f8d0e56121719b1a3e8cdecf2.

This could compile if `target_os = "macos"` was a subset of `target_family = "unix"`.

Specifically, two extra relations can be defined:
- `cfg(target_os = "windows")` ⊆ `cfg(target_family = "windows")`.
- `cfg(target_os = <unix-os>)` ⊆ `cfg(target_family = "unix")`, where `<unix-os>` is any of
  `["freebsd", "linux", "netbsd", "redox", "illumos", "fuchsia", "emscripten", "android", "ios",
  "macos", "solaris"]`. This list needs to be updated if a new `unix` OS is supported by `rustc`'s
  official target list. This would make the first example compile.

_Note:_ The contrapositive of these relations is also true.

Also, `target_family` is currently defined as not having mutually exclusive elements. This is
because `target_family = "wasm"` is not mutually exclusive with other target families. But,
`target_family = "unix"` could be defined as mutually exclusive with `target_family = "windows"` to
increase usability. By extension, `target_family = "windows"` would now be mutually exclusive with
`target_os = "linux"`, for example.

_Note:_ More relations could be defined, for example `target_feature = "neon"` ⊆ `target_arch =
"arm"`. With this however, things start to get complicated.

> ## Lint against useless target-specific tables
> 
> If a package has:
> ```toml
> [package]
> name = "example"
> # ...
> supported-targets = 'cfg(target_os = "linux")'
> 
> [target.'cfg(target_os = "windows")'.dependencies]
> # ...
> ```
> A lint could be added to highlight the fact that the `[target]` table is useless.

Reviewer: ```suggestion
## Lint against unused target-specific tables

If a package has:
```toml
[package]
name = "example"
# ...
supported-targets = 'cfg(target_os = "linux")'

[target.'cfg(target_os = "windows")'.dependencies]
# ...
```
A lint could be added to highlight the fact that the `[target]` table is unused.
```
a little less charged and aligning with similar lints

## `supported-targets` at the cargo-target level

The `supported-targets` field could also be added at the cargo-target level to have more
fine-grained control over which targets a cargo-target supports. This would function similarly to
the `edition` field, which is available at both the package and the cargo-target level. The
`supported-targets` of a cargo-target would most likely need to be a subset of the package's

> `supported-targets`.

Reviewer: Would be good to link out to 
https://github.com/rust-lang/cargo/issues/12260
https://github.com/rust-lang/cargo/issues/9208

for context that would need to be taken into account with this

Author: Added in https://github.com/rust-lang/rfcs/pull/3759/commits/4256e8234770e4f7a727e961d31b64eb389861b0.

specify platform requirements of a package

When working on a project with packages that only build on certain platforms, users cannot run Cargo commands across the entire workspace (e.g. `cargo test --workspace`) but must individually select packages that only work on the specific platform (e.g. `cargo test --workspace --exclude firmware`).  This extends to CI with people wanting to write matrix jobs but have to hand maintain the list of packages for each platform in the matrix.

## Long-term Motivations

This RFC unblocks further work to improve platform-specific packages.  While these problems are important, solving them has been left to  [future
possibilities](#future-possibilities) to deliver an MVP we can then build on.

### More specific error messages

The error message when a library has platform-specific features, like requiring atomics, is about parts of `std` missing which could be for one of several reasons. Some of these problems won't be found until you've built or tested your project on one of these platforms. Like with #2495, if library authors could provide this information to Cargo, developers can get an improved error message under any circumstance.

> ### Dependency Management 
> 
> Once it is known that a package will only ever build for a subset of targets, it opens the door for
> more advanced control over dependencies. For example, transitive dependencies declared under a
> `[target.**.dependencies]` table could be excluded from `Cargo.lock` if the dependent's
> `supported-targets` are mutually exclusive with the target preconditions under which the
> dependencies are included. This is especially relevant to areas such as WebAssembly and embedded
> programming, where one usually supports only a few specific targets. Currently, auditing and
> vendoring is tedious because dependencies under `[target.**.dependencies]` tables always make their
> way in the dependency tree, even though they may not actually be used.

Reviewer: ```suggestion
### Include fewer packages with `cargo vendor`

`Cargo.lock`, and by extension, `cargo vendor`, must assume that a package may be built on any platform that has or will exist.  This means that if a transitive dependency pulls in Windows-specific dependencies, `cargo vendor` will include them when run on a Linux-only application.  Being able to tell `cargo vendor` what platforms to care about can reduce the space used in a repo and reduce churn.

### Dependency Management 

Likewise, today users either need to audit dependencies irrelevant for the platforms they target or filter these out somehow.  By providing first-class support for specifying what platform features a package requires, audit tools can consolidate on that for narrowing down the list of what dependencies to audit.
```

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> The `supported-targets` field can be added to `Cargo.toml` under the `[package]` table.

Reviewer: Does this field support workspace inheritance?

Reviewer: Note: fully specifying the field should probably belong in the Reference section.  Maybe write it as if it is the documentation you'd put in https://doc.rust-lang.org/cargo/reference/manifest.html

Author: I added some details to the reference section, including that the field supports workspace inheritance.

experience is enhanced by raising an error that fails compilation when the supported targets of a
package are not satisfied by the selected target.

This feature should be used when a package clearly does not support all targets. For example:
`io-uring` requires `cfg(target_os = "linux")`, `gloo` requires `cfg(target_family = "wasm")`, and
`riscv` requires `cfg(any(target_arch = "riscv32", target_arch = "riscv64"))`.

This feature increases cargo's knowledge of a package. For example, when working in a workspace
where some packages are for a platform with `target_os = "none"`, and some others are tools that
require a desktop OS, using `supported-targets` makes `cargo <command>` ignore packages which have
`supported-targets` that are not satisfied by the selected target.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

When a `cargo` build command (e.g. `check`, `build`, `run`, `clippy`) is run, it checks that the
selected target satisfies the `supported-targets` of the package being built. If it does not, the
package is skipped or an error is raised, depending on how [`cargo` was invoked](ignoring-builds).

> However, `supported-targets` are _not_ checked if the cargo command does not require compilation
> (e.g., `cargo fmt`). This field is only for local development, and is removed from `Cargo.toml`
> when publishing a crate. It could be published to give more information to `docs.rs`, `crates.io`,
> and `cargo` itself, but this is left as a [future possibility](#future-possibilities).

Reviewer: ```suggestion
However, `supported-targets` is _only_ checked for commands that take a `--target` option and does not affect other commands (e.g., `cargo fmt`).

As this field is limited to local development, `cargo package` / `cargo publish` will strip it from `Cargo.toml`.
Including the field in the `.crate` file is left as a [future possibility](#future-possibilities) for now.
```

- Feature Name: `supported-targets`
- Start Date: 2025-01-08
- Pre-RFC: [Rust
  internals](https://internals.rust-lang.org/t/pre-rfc-allow-packages-to-specify-a-set-of-supported-targets/21979)
- RFC PR: [rust-lang/rfcs#3759](https://github.com/rust-lang/rfcs/pull/3759)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

The word _target_ is extensively used in this document. The
[glossary](https://doc.rust-lang.org/cargo/appendix/glossary.html#target) defines its many meanings.
Here, _target_ refers to the "Target Architecture" for which a package is built. Otherwise, the
terms "cargo-target" and "target-triple" are used in accordance with their definitions in the
glossary.

# Summary

The addition of `supported-targets` to `Cargo.toml`. This field is a `cfg` string that restricts the
set of targets which a package supports. Packages can only be built for targets that satisfy their
`supported-targets`.

# Motivation

> Some packages do not support every possible `rustc` target. Currently, there is no way to formally
> specify which targets a package does, or does not support.

Reviewer: ```suggestion
Some packages rely on features or behavior that is not available on every platform `rustc` can target. Currently, there is no way to formally
specify platform requirements of a package
```

- Feature Name: `supported-targets`
- Start Date: 2025-01-08
- Pre-RFC: [Rust
  internals](https://internals.rust-lang.org/t/pre-rfc-allow-packages-to-specify-a-set-of-supported-targets/21979)
- RFC PR: [rust-lang/rfcs#3759](https://github.com/rust-lang/rfcs/pull/3759)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

The word _target_ is extensively used in this document. The
[glossary](https://doc.rust-lang.org/cargo/appendix/glossary.html#target) defines its many meanings.
Here, _target_ refers to the "Target Architecture" for which a package is built. Otherwise, the
terms "cargo-target" and "target-triple" are used in accordance with their definitions in the
glossary.

# Summary

The addition of `supported-targets` to `Cargo.toml`. This field is a `cfg` string that restricts the
set of targets which a package supports. Packages can only be built for targets that satisfy their
`supported-targets`.

# Motivation

Some packages do not support every possible `rustc` target. Currently, there is no way to formally

> specify which targets a package does, or does not support.

Reviewer: Let's try to be careful with the using the word "support" as its causing people to read more into it than what is intended with this RFC.

Author: "suitable"?

I have a limited english vocabulary so I have a hard time finding something better.

Reviewer: I gave a suggestion for this instance at https://github.com/rust-lang/rfcs/pull/3759/files#r1919180344

> This feature enhances developer experience when working in workspaces containing packages designed
> for many different targets. Commands run on a workspace ignore packages that don't support the
> selected target.

Reviewer: This is a description and not a motivation
```suggestion
When working on a project with packages that only build on certain platforms, users cannot run Cargo commands across the entire workspace (e.g. `cargo test --workspace`) but must individually select packages that only work on the specific platform (e.g. `cargo test --workspace --exclude firmware`).  This extends to CI with people wanting to write matrix jobs but have to hand maintain the list of packages for each platform in the matrix.
```

## Long-term Motivations

> _These do not motivate the RFC itself, rather they motivate the [future
> possibilities](#future-possibilities) section._

Reviewer: I disagree with this.  I'm unsure if we would approve this RFC without some of these other motivations.

```suggestion
This RFC unblocks further work to improve platform-specific packages.  While these problems are important, solving them has been left to  [future
possibilities](#future-possibilities) to deliver an MVP we can then build on.
```

- Feature Name: `supported-targets`
- Start Date: 2025-01-08
- Pre-RFC: [Rust internals](https://internals.rust-lang.org/t/pre-rfc-allow-packages-to-specify-a-set-of-supported-targets/21979)
- RFC PR: [rust-lang/rfcs#3759](https://github.com/rust-lang/rfcs/pull/3759)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

The word _target_ is extensively used in this document.
The [glossary](https://doc.rust-lang.org/cargo/appendix/glossary.html#target) defines its many meanings.

> Here, _target_ refers to the "Target Architeture" for which a package is built. Otherwise, the terms

Reviewer 4: ```suggestion
Here, _target_ refers to the "Target Architecture" for which a package is built. Otherwise, the terms
```

# Summary

The addition of `supported-targets` to `Cargo.toml`.
This field is an array of `target-triple`/`cfg` specifications that restricts the set of targets which
a package supports. Packages must meet the `supported-targets` of their
dependencies, and they can only be built for targets that satisfy their `supported-targets`.

# Motivation

Some packages do not support every possible `rustc` target. Currently, there is no way to formally
specify which targets a package does, or does not support.

### Developer Experience

Trying to depend on a crate that does not support one's target often produces cryptic build errors,
or worse, fails at runtime. Being able to specify which targets are supported ensures that unsupported
targets cannot build the crate, and also makes build errors specific.

This feature also enhances developer experience when working in workspaces containing packages designed for
many different targets. Commands run on a workspace ignore packages that don't support the selected target.

### Cross Compilation 

Once it is known that a package will only ever build for a subset of targets, it opens
the door for more advanced control over dependencies.

> For example, transient dependencies declared under a `[target.**.dependencies]` table are

Reviewer 4: ```suggestion
For example, transitive dependencies declared under a `[target.**.dependencies]` table are
```

(I'm assuming this is what was meant.)

excluded from `Cargo.lock` if the dependent's `supported-targets` is mutually exclusive with
the target preconditions under which the dependencies are included.
This is especially relevant to areas such as WebAssembly and embedded programming,
where one usually supports only a few specific targets. Currently, auditing and vendoring
is tedious because dependencies under `[target.**.dependencies]` tables always make their way
in the dependency tree, even though they may not actually be used.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

The `supported-targets` field can be added to `Cargo.toml` under the `[package]` table.

This field consists of an array of strings, where each string is an explicit target-triple or a `cfg` specification
(as for the `[target.'cfg(**)']` table). The supported `cfg` syntax is the same as the one for
[platform-specific dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies)

> (i.e., `cfg(test)`, `cfg(debug_assertions)`, and `cfg(proc_macro)` are not supported).

Reviewer 5: How does this interact with crate features and optional dependencies? In some cases, these impose additional platform requirements, e.g., because the `alloc::sync` types don't exist on platforms without atomics. How could this be modeled, if at all? Platform-specific dependencies aren't allowed to mention `cfg(feature = "...")`, but that would be required to express e.g., "if the `std` feature is enabled, I require `cfg(target_has_atomic="ptr")`".

Reviewer 6: > the alloc::sync types don't exist on platforms without atomics

It seems pretty bad to express that via a mechanism such as the one proposed in this RFC. Instead, that should be expressed by saying "I need atomics", and then rustc knows what that means in terms of which targets can and cannot be supported.

Reviewer 6: I don't know, I am still struggling to understand the motivation for this RFC. ;) Presumably it would have nicer error messages or so?

Reviewer 5: I agree that the part of the RFC that lets package authors list specific target tuples is a bad idea. Many others raised similar concerns. I’m convinced it should only be expressed in terms of `cfg` values. And my point in this specific review sub-thread is that these requirements will often depend on enabled crate features.

Author: I have added a small note on this in _Future Possibilities_ in https://github.com/rust-lang/rfcs/pull/3759/commits/f3843cf85b1a8610cdeee1d8f2d3a9cfa4e4becd.

If a selected target satisfies any entry of the `supported-targets` list, then the package can be built for that target.

__For example:__
```toml
[package]
name = "hello_cargo"
version = "0.1.0"
edition = "2021"
supported-targets = [

>     "wasm32-unknown-unknown",

Reviewer 7: I think that only the cfg version should be allowed, and explicit target tuples should not. It makes it way to easy to accidentally make a crate overly restrictive. The cfg form ensures that the target only restricts what it truly needs. I can't imagine a crate that can only support a precise single target (on fact, this example is a great example of this, as the wasm target name is likely to change in the future).

Reviewer 7: I think doing any matching on full target tuples is a mistake. It will be overly restrictive in almost all cases, leading to both needless incompatibilities today and bad future compatibility in the future.
I don't think we should have consistency for mistakes.
This entire RFC, while useful, presents a serious risk to needless incompatibilities for niche and future targets, and this should be mitigated as much as possible.
Only supporting cfgs and therefore steering users away from being overly restrictive is a good way to do that.

Reviewer 8: Embedded packages for dealing with a device might easily want precise target names.

I don't think that 99% of the ecosystem should ever use this feature, but some crates really are intended for that narrow of a purpose.

Reviewer 5: If you mean code that's specific to one SoC/board, target tuples are both too narrow and too broad to work well for that purpose (someone might still *want* it, but it seems ill-advised). There can be multiple "correct" tuples for the same device. The most basic example is custom target specs: while unstable, it's useful to be able to take an existing built-in target and slightly tweak it. But even built-in target are not safe:

- As the RFC also discusses, targets can be removed or renamed.
- A single target may be split up into several, e.g., enable more fine grained distinctions that didn't matter before.
- Multiple targets may be merged because there's a better way to handle their differences, e.g., soft/hard float via the "target modifier" concept proposed in https://github.com/rust-lang/rfcs/pull/3716 plus stabilization of `-Zbuild-std`.

(Of course, all of this also causes problems for the existing `[target.a-b-c-d.dependencies]` tables, but as @Noratrieb says that's no reason to double down on it.)

Reviewer 9: @Lokathor is there a reason you can't make use of one of `target_vendor`, `target_env`, `target_os`, `target_abi`? 

We've previously rejected part of a similar RFC that allowed specifying full target names for the same reasons as presented here.

Reviewer 8: I suppose it's possible.

Reviewer 2: > My original proposal was to only support target tuples because it allowed for very precise trimming of Cargo.lock / cargo vendor.

That would be very incorrect trimming, also, unfortunately. Let us imagine this was introduced earlier in Rust's lifetime and it was used for a package that only supported Windows. A package would have to declare it only allows `["x86_64-pc-windows-msvc", "i686-pc-windows-msvc", "x86_64-pc-windows-gnu", "i686-pc-windows-gnu"]`.

Then suddenly the gnullvm, arm64ec, and aarch64 targets would utterly explode that, even if it worked fine on all of those without changes.

Author: I would be happy with only supporting `cfg`s. In this case we could have the field be a single string, e.g.:
```toml
supported-targets = 'cfg(any(target_family = "unix", target_family = "wasm"))'
```
We could even technically omit the "cfg" prefix (not a good idea imo, this would be confusing).

    'cfg(target_os = "linux")',
    'cfg(target_os = "macos")'
]
```
Here, only targets satisfying: the `wasm32-unknown-unknown` target, __or__ the `linux` OS, __or__ the `macos` OS,
are allowed to build the package. If no `supported-targets` was specified, then any target would be allowed.

User experience is enhanced by raising an error that fails compilation when the supported targets
of a package are not satisfied by the selected target. A package's `supported-targets` must be a subset

> of its dependencies' `supported-targets`, otherwise the build also fails.

Reviewer 4: This seems helpful, but at the same time, it may prove annoying to have to copy these across. And it may be non-trivial to enforce: consider the case of multiple target-specific dependencies that each have different supported-targets, and showing that the result is equivalent.

I think we should downgrade this to a lint, and say that it's best-effort, not mandatory.

Reviewer 8: If it's just a lint then unsafe coders can't depend on this, and they will still need to just use a `compile_error!` or something if they're really trying to avoid unsoundness.

Reviewer: This is effectively saying that if you have a dependency that can't build on a target you claim to support, the build will fail and you should instead move it to a `target.*.dependencies` table.

imo that seems like something that should be a hard error to me.

I could see loosening the restriction on

> When supported-targets is not specified, any target is accepted, so all dependencies must support all targets.

To be "we don't check", like `package.rust-version`, requiring `supported-targets = ["cfg(true)"]` to get the checking.

Author: I totally agree with @epage here.

Using the same logic as `package.rust-version` is also something I have not thought of but is a great idea.

Author: A problem I could see however is that if a package `foo` does not use `supported-targets` but one of its dependencies does, then when depending on `foo` errors of incompatible targets can be hard to solve since they come from transitive dependencies.

Reviewer: Not quiet sure the problem.  When building `foo`, you are building for a specific target and the error would be for that target.  With `package.rust-version`, we check what all packages aren't compatible with the current toolchain and provide a single error message, see https://github.com/rust-lang/cargo/blob/9589831f61a8259919e64c6d68c1a36efc6efd20/src/cargo/ops/cargo_compile/mod.rs#L494-L543

Reviewer 4: > If it's just a lint then unsafe coders can't depend on this, and they will still need to just use a `compile_error!` or something if they're really trying to avoid unsoundness.

supported-targets would still be enforced (modulo some kind of force option); I'm talking about softening the enforcement that a crate's supported-targets is a subset of its dependencies.

Reviewer: FYI this has moved to a Future Possibility

> of its dependencies' `supported-targets`, otherwise the build also fails.

Reviewer 7: I don't think this actually works, if I understand this correctly. A package can very well support more targets than its dependencies, by only conditionally using these dependencies (cfged at compile time).

Author: [Here](https://github.com/rust-lang/rfcs/pull/3759/files#diff-7ca74978b782c1ea19b9cafec0cd71f3936a4791ac30656f361896ee217d391cR125). They just need to be moved to platform specific dependencies.

When `supported-targets` is not specified, any target is accepted, so all dependencies must support
all targets.

This feature should be used when a package clearly does not support all targets. For example: `io-uring`
requires `cfg(target_os = "linux")`, `gloo` requires `cfg(target_family = "wasm")`, and
`riscv` requires `cfg(target_arch = "riscv32")` or `cfg(target_arch = "riscv64")`.

This feature should also be used to increase cargo's knowledge of a package. For example,
when working in a workspace where some packages compile with `#[no_std]` and `target_os = "none"`, 
and some others are tools that require a desktop OS, using `supported-targets` makes
`cargo <command>` ignore packages which have `supported-targets` that are not satisfied
by the selected target.

> 

Reviewer 4: ```suggestion

Cargo's documentation should give clear guidance for when to use this field, and should not suggest using it by default. In particular, we should steer users to use this when they have good reason to believe the crate will not compile or work as expected (e.g. because it uses target-specific APIs), and not use it merely for "I haven't personally tested this on other targets".

```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> When `cargo` is run on a package, it checks that the selected target satisfies the `supported-targets`

Reviewer 4: ```suggestion
When a `cargo` build command (e.g. `check`, `build`, `run`, `clippy`) is run on a package, it checks that the selected target satisfies the `supported-targets`
```

> of the package. If it does not, an error is raised and the build fails. However, `supported-targets`

Reviewer: Why is this erroring, rather than following the skipping logic of `required-features`?  

This makes `cargo check --workspace` impossible to use and imo it would be good to be consistent with `required-features`

Author: This is just poorly written by me, the behavior is detailed starting [line 274](https://github.com/rust-lang/rfcs/pull/3759/files#diff-7ca74978b782c1ea19b9cafec0cd71f3936a4791ac30656f361896ee217d391cR274). What you described, `cargo check --workspace` would indeed ignore packages instead of erroring.

I'll make sure to correct that.

Reviewer 5: Another part of the RFC, titled "Ignoring builds for unsupported targets," says:

> When the target used is not supported by the package being built, the package will either be skipped or an error will be raised, depending on how cargo was invoked. If cargo is invoked in a workspace or virtual workspace without specifying a specific package, then cargo skips the package. If a specific package is specified using --package, or if cargo is invoked on a single package, then an error is raised.

I'm not sure how this would be implemented in Cargo but at first glance it seems reasonably consistent with how `required-features` works. Trying to build a *specific* binary/example that has unmet `required-features` also errors (see https://github.com/rust-lang/cargo/issues/4663). I think there would be some subtle differences due to one being per-cargo-target and the other being per-package (e.g., `cargo install` skips the binaries that have unmet `required-features`, even if that ends up installing nothing).

is _not_ checked if the cargo command does not require compilation (`cargo doc`, `cargo fmt`, etc.).

> For example, `cargo doc` can be run on a host that the package does not support.

Reviewer: Is this true for `cargo doc`? Documentation changes based on the `--target` used (hence why docs.rs gives control over it) and I'm assuming its using the cross-compilation logic.

Author: I did not consider that properly, and I did not know that docs.rs gives control over it.

Then I believe one could build the docs by running `cargo doc --target <target>` (assuming one has the toolchain installed) even if the host target is not supported. So this solves [@Lokathor's issue](https://github.com/rust-lang/rfcs/pull/3759#issuecomment-2578515182).

I'll remove this.

Reviewer: See https://docs.rs/about/metadata

## Compatibility of `[dependencies]`

The set of `supported-targets` of a package must be a subset of the `supported-targets` of its
`[dependencies]`. If the crate itself has no `supported-targets` specified,
then all dependencies must support all targets.

> If a dependency does not respect this requirement (if it is not compatible), an error is raised and the build fails.

Reviewer 10: As per some non-threaded comments on this PR by @CryZe and @ahicks92: this feature has the risk that package authors will list a *too narrow* set of `supported-targets`. They can already do this with a conditional `compile_error!`, but I am particularly concerned that **they might do so on the optimistic principle “filling in all of the metadata is good, right?”,** whereas if the manifest field did *not* exist, they wouldn’t have thought to write a `compile_error!`.

I think that, in order to reduce the risk of gratuitous breakage that requires users of a not-eagerly-maintained package to fork/vendor it or find an alternative, this should be *advisory* for dependents: it does not force a build to fail unconditionally, but causes a lint that might error by default but can be opted out of in some way (perhaps as a field in the `[dependencies]` table).

(By “not eagerly maintained”, I mean: the code is generally good quality, but it takes weeks/months to get a PR in upstream, or for a release to happen after the PR. I think that it is an important virtue of Rust that you can usually successfully use such packages, rather than various kinds of churn making it hard.)

In order to support the “it’s UB to run this code on (not) this target” use case, there can be a way to require it to be a hard error (or at least, not bypassable by dependents, only by the build command configuration). But that should be opt-in, so that the feature can also be used for the softer use case of offering **machine-readable requirements documentation** and avoiding messes of “item/symbol not found” error messages.

Reviewer 5: For the "it's UB to run this code on (not) this target" use cases, the existing tools of conditional `compile_error!(...)` / `const { assert!(...); }` seems better than overloading the `supported-targets` field. Putting each assumption right next to the `unsafe` code that relies on it means:

- It's easier to audit the code because it's more localized -- even if the safety comment on a block justifies something by reference to Cargo.toml, I as reviewer have to check that.
- It's harder to forget about reviewing/updating the code when changing the required-targets definition. You could put a comment in the relevant part of the manifest, but that's further removed from the relevant code so it's easier for the comment to go out of sync.
- If I depend on a library that has (in my view) an overly restrictive `supported-targets`, I can try to work around this by patching it locally. Editing the manifest is easy, bypasses any "enforcement" by Cargo, and makes it easy to miss that some `unsafe` code in the guts of the library actually relies on it for soundness. Having to dive into the source code and remove static assertions next to `unsafe` blocks is higher friction, which is good when soundness hinges on it.

Reviewer 11: If the opt-out is encoded in Cargo.toml, then Cargo can see that when generating the lockfile, right? So CLI flags *do* make it impossible to trim the lockfile, but explicit override (similar to how default / non-default features work in principle) seems doable?

Reviewer: Cargo resolves `Cargo.lock` assuming all features of workspace members may be activated.  It only trims features from transitive dependencies where there is no way to activate them through the command-line

An `--ignore-supported-targets` flag is a transient state and the lockfile generally would need to take into account all transient states in its creation so it doesn't change based on how Cargo commands invoke it.

In the past, I have toyed with the idea that the lockfile could remember some types of resolution inputs to re-apply them in the future.  This is a very dodgy area that would need a lot of detail to figure out and is questionable whether it would ever get approved.  This is the type of thing that could slow down or derail the RFC and would ideally be split out into a future possibility to make this RFC go more smoothly.

Reviewer 4: It sounds like in theory we could provide a non-transient force/override option in the manifest, which could then disable trimming the dependencies. But :+1: for not blocking on that in any case.

Reviewer 12: I share the concern that people might write target requirements that are too strict, or that they might write target requirements that were correct _at the time_ but things have changed and the crate hasn't been updated.  I want to suggest that:

- When _directly_ building a package (i.e. not as a dependency), attempting to build for a target that it declares it doesn't support should trigger a deny-by-default lint.
- When a package is being built as a dependency, attempting to build for a target that it declares it doesn't support should trigger an _allow_-by-default lint.

This is analogous to honoring `Cargo.lock` for the package being directly built but not for its dependencies.  If, hypothetically speaking, the `io_uring` API becomes available on FreeBSD at some time in the future, but the maintainer of some `io_uring` interface package neglects to add FreeBSD to the list of supported targets for that package, binary crates that use that package can still be built for FreeBSD despite the exclusion.

Reviewer: 1. I think we need to keep in mind that the framing of this RFC can be improved which affects the conversation, e.g. https://github.com/rust-lang/rfcs/pull/3759#discussion_r1909208702
2. We need to make sure we are clear in communicating how this should be used, see https://github.com/rust-lang/rfcs/pull/3759#discussion_r1909413295

That said, there are likely still times to override.  I think we're being a bit limited in exploring our options

The discussed CLI override
- Affect the entire dependency tree
- Prevents `Cargo.lock` trimming
- Every dependent must do it

Lint overrides
- See CLI

A single manifest  override (bespoke version of lint override)
- See CLI

Switching to mutable metadata, see https://github.com/rust-lang/rfcs/pull/3759#issuecomment-2581681334
- Whole new process that would need to be designed and explored

Other options
- Patching (existing)
  - Requires copying the entire repo somewhere (maybe `cargo override` can help?)
  - Every dependent must do it
- Unidiff patch (proposed)
  - More limited in what needs to be in your repo
  - Has some design questions that need working out
  - Every dependent must do it
- Manifest override on specific dependencies
  - Every dependent must do it

Reviewer: Another idea I thought of: punt on incompatible-dependency error reporting.

Error reporting is a non-essential part of this RFC.  The main reason for it is to improve the error messages when you build something on an incompatible platform (though I think some have brought up `unsafe` assumptions which is slightly more important).  iirc this wasn't even one of the original motivations for the requesting issue (build-target filtering and `Cargo.lock` filtering are).

By delaying error reporting support,
- We remove a controversial aspect of this RFC that can make it easier to merge
- Set the tone for how this feature should be used when error reporting gets added in by the fact that the common use of this feature is for build-target filtering and `Cargo.lock` pruning.

We can switch on error reporting in an edition.  We could only respect a dependency's `supported-targets` if its edition is new enough or we could only pay attention to the current package's edition.

I'd suggest we strip `supported-targets` on publish.  It is not needed in dependencies for pruning as they'll either be compatible or be behind a `target.*dependencies` to limit them to when they are compatible.  While it could be helpful for documentation purposes or for when error reporting is added, by stripping the field we make it clear that this feature is only being used for local development for now and avoids us backfilling the registry with controversial choices.

Reviewer: The RFC scope has been narrowed down to just package/build-target filtering when running Cargo commands and the field is stripped on publish to prevent it from being misused.  Erroring has been moved to a future possibility and we can discuss it further when/if someone proposes that in the future.

Author: > The RFC scope has been narrowed down to just package/build-target filtering when running Cargo commands and the field is stripped on publish to prevent it from being misused. Erroring has been moved to a future possibility and we can discuss it further when/if someone proposes that in the future.

I will resolve the conversation.

If this was not enforced, a package could support targets that are not supported by its dependencies,
which does not make sense.

## Compatibility of `[dev-dependencies]`

`[dev-dependencies]` are checked the using the same method as regular `[dependencies]`. That is, the package's
`supported-targets` must be a subset of every `[dev-dependencies]`'s `supported-targets`. The rationale is
that an example, test, or benchmark has access to the package's library and binaries, and so it must respect the

> `supported-targets` of the package.

Reviewer 4: I don't think this should be a requirement. It's sometimes reasonable for tests to only run on some targets.

Reviewer: But we'll try to build the dev-dependency on an unsupported target, independent of what `cfg` is on the test, which will fail.  The dev-dependency should be moved to a `target.*.dev-dependencies` table, bypassing this.

Author: > The dev-dependency should be moved to a `target.*.dev-dependencies` table, bypassing this.

The RFC does not do a good job of "teaching" how to use this feature (Could be added to guide-level explanation), but this is what I have in mind.

_intersection_ of the package's supported-targets, and the target conditions it is declared under.
For example:
```toml
[package]
# ...
supported-targets = ['cfg(target_os = "linux")']

[target.'cfg(target_pointer_width = "64")'.dependencies]
foo = "0.1.0"
```
Here, it suffices for `foo` to support `cfg(all(target_os = "linux", target_pointer_width = "64"))`.

## Artifact dependencies

If an artifact dependency has a `target` field, then the dependency is not checked against the
package's `supported-targets`. However, the selected `target` for the dependency must be compatible with
the dependency's `supported-targets`, or else an error is raised. If the artifact dependency does not have a
`target` field, then it is checked against the package's `supported-targets`, like any other dependency.

## Comparing `supported-targets`

When comparing two `supported-targets` lists, it is necessary to know if one is a _subset_ of the other,
or if both are _mutually exclusive_. To proceed, both lists are flattened to
the same representation, and they are then compared. This process is done internally, and does not
affect the `Cargo.toml` file.

> ### Flattening `not`, `any`, and `all` in `cfg` specifications

Reviewer 5: Does this section just describe conversion to disjunctive normal form? The array of strings is the top-level logical OR, and each entry is the AND of (possibly negated) atoms. If this correct, it would be helpful to make this explicit because it saves anyone who has encountered DNF before some reverse-engineering. This also makes it easier to see that this conversion can cause exponential blowup in the worst case (not a fatal problem, since dependency resolution is also NP-hard, but good to keep in mind).

Author: I have no clue what that is, I just decided on a representation that "felt right" (I have very little academic knowledge on SAT solving). I used this representation because I thought that no-one would ever have crazy nested `cfg` specifications. I'll read up on it and link it there.

Author: I believe it is indeed DNF, I added a note about it in https://github.com/rust-lang/rfcs/pull/3759/commits/52a7ebbd3630007d43b19a8fe26c97d8a7319ed4.

## Behavior with unknown entries (and custom targets)

Just as `[target.my-custom-target.dependencies]` is allowed by `cargo`, `supported-targets` can contain
unknown entries. This is important because users may have different `rustc` versions, and the set of
official `rustc` targets is unstable; Targets can change name or be removed. Also, developers 
may want to support their custom target.

To determine if an entry in `supported-targets` is a target name or a `cfg` specification, the
same mechanism as for `[target.'cfg(..)']` is used (using
[`cargo-platform`](https://docs.rs/cargo-platform/latest/cargo_platform/index.html)).

## Ignoring builds for unsupported targets

When the target used is not supported by the package being built, the package will either be
skipped or an error will be raised, depending on how `cargo` was invoked. If cargo is invoked
in a workspace or virtual workspace without specifying a specific package, then `cargo` skips the package.
If a specific package is specified using `--package`, or if `cargo` is invoked on a single package,
then an error is raised.

## Eliminating unused dependencies from `Cargo.lock`

> A package's dependencies may themselves have `[target.'cfg(..)'.dependencies]` tables, which may never be used because of the

Reviewer: As this affects dependency resolution, we'll also need to discuss the relevant changes to the Index.

For the existing schema, see https://doc.rust-lang.org/cargo/reference/registry-index.html#json-schema

- Performance: this is a lot of extra calls to rustc.
    Hopefully these are all compatible with the rustc cache, so they won't make things too bad.
- This feature must be learned by those wanting to use dependencies with `supported-targets` specified.
- This is the first step towards a target aware `cargo`, which may increase `cargo`'s complexity, and bring
    more feature requests along these lines.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

## Format

The list of strings format was chosen because of its simplicity and expressiveness. Other formats can also be considered:

Using the `[target]` table, for example:
```toml
[target.'cfg(target_os = "linux")']
supported = true
```
If the list of supported targets is long (should it ever be?), then the `Cargo.toml` file becomes very verbose
as well.

A `[suppported]` table, with `arch = ["<arch>", ...]`, `os = ["<os>", ...]`, `target = ["<target>", ...]`, etc.
This is more verbose, complex to implement, learn, and remember. It is also not obvious how `not` and `all`
could be represented in this format. For example:
```toml
[supported]
os = ["linux", "macos"]
arch = ["x86_64"]

> ```

Reviewer: This leaves off why an array is used instead of just a string

Author: Added in https://github.com/rust-lang/rfcs/pull/3759/commits/e19cbf6a5f54855d24b03d8836d753920330e5e2.

## Naming

Some other names for this field can be considered:

- `required-targets`. Pro: it matches with the naming of `required-features`. Con: `required-features` is a list of features
    that must _all_ be enabled (conjunction), whereas `supported-targets` is a list of targets

>     where _any_ is allowed (disjunction).

Reviewer: For me the important precedence is that the field means "skip if the qualification is not met" and for that reason I favor using this name.

Reviewer: Apparently, this is different than `required-features` as this errors, rather than skips.

I raised this at https://github.com/rust-lang/rfcs/pull/3759/files#r1909208702

Reviewer: I also feel like using the word "support" carries a lot of unnecessary connotations that complicate the conversation.  For example, with MSRV, the Cargo team has been leaning in the direction that "support" is an active process that gets tested.  However, applying that here would lead to people over-constraining their targets.

Author: The difference with this and `required-features` is that `supported-targets` is at the package level, not at the cargo-target level.

So if I run `cargo build` in a package with a binary `foo` which does _not_ have its `required-features` met, then `cargo` can still possibly "do work" if there is another cargo-target which has its `required-features` met (for instance a library cannot have any `required-features`).

However if the `supported-targets` are not met for a package, then there is no chance of cargo doing compilation work for that package. That is why the packages are skipped when in a workspace, but an error is raised in a single package. This is just how if I ran `cargo build --bin foo` in the previous example, then `cargo` errors instead of skipping.

Author: I do not have a preference for the name, I kept it as is to not confuse people who had read the Pre-RFC. I would not be against changing it.

Reviewer: I don't think this being at the package or build-target level makes much of a difference.  The RFC is starting at the package level and has build-target as a future possibility.  In that case, we can treat this as the package is providing the default for all build-targets like `package.edition`

- `targets`. As in "this package _targets_ ...". Pro: Concise. Con: Ambiguous, and could be confused with the `target` table.

## Package scope vs. cargo-target scope

> The `supported-targets` field is placed at the package level, and not at the cargo-target
> level (i.e., under, `[lib]`, `[[bin]]`, etc.). A rationale is given for why the cargo-target level is not used.

Reviewer: ```suggestion
The `supported-targets` field is placed at the package level, and not at the cargo-target
level (i.e., under, `[lib]`, `[[bin]]`, etc.)
```
This is a bit redundant as the intention of having a section like this is to provide the rationale.  Having it here makes it read awkwardly to me, including that it implies its going to say something different from the norm but it doesn't.

here along with their drawbacks.

### Using wildcards

Instead of using `cfg` specifications, one could use wildcards (e.g., `x86_64-*-linux-*`).
This is much simpler to implement, target-triples are syntactically checked for a match instead
of solving set relations for `cfg`. However, this is not as expressive as `cfg`, and does not correctly
represent the semantics of target triples. For example, supporting `target_family = "unix"` would
require an annoyingly long list of wildcard patterns. Things like `target_pointer_width = "32"` are
even harder to represent, and things like `target_feature = "avx"` are basically not representable.
Also, this is new syntax not currently used by cargo.

### Allowing only target triples

This is an even stricter version of the above. Being even simpler to implement, this alternative
may not be expressive enough for the common use case. Packages rarely support specific target triples,
rather they support/require specific target attributes. What would likely happen is that packages
would copy and paste the target triple list matching their requirements from somewhere or someone else.

> Every time a new target with the same attribute is added, the whole ecosystem would have to be updated.

Reviewer: This should call out the reason why it was originally proposed
- Easier, more exact relations logic
- Easier for the resolver to leverage this information

Author: Added in https://github.com/rust-lang/rfcs/pull/3759/commits/565bf85e004f04ffa0188b822415a0fc5892c9f4.

to be built for the same target as the package. `supported-targets` supersedes `force-target` because instead of enforcing
a single target, it enforces a set of targets.

I am not aware of a package manager that solves this issue like in this RFC. In other system level languages,
vendoring dependencies is a common practice, and the user would be responsible for ensuring that the dependencies
are compatible with the target. For interpreted languages, this is a non-issue because any platform being able
to run the interpreter can run the package.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- What if one wants to exclude a single target-triple? Groups can be excluded with `cfg(not(..))`, but there
	is currently no way of excluding specific targets (Would anyone ever require this?).
- Some crates will inevitably have target requirements that are too strict, how
    do we make users bypass the error (probably with some `--ignore-**` flag)? Do we want to allow this?
- Should we solve for this during dependency version resolution? (the current rationale is that we do not want

> 	targets to affect package version resolution).

Reviewer: As posted at https://github.com/rust-lang/rfcs/pull/3759#discussion_r1909402331

> one of the intended use cases is for this to trim down `Cargo.lock` / `vendor/`. If its advisory only, then we can't do that anymore.

- `cfg(target_os = "windows")` ⊆ `cfg(target_family = "windows")`.
- `cfg(target_os = <unix-os>)` ⊆ `cfg(target_family = "unix")`, where `<unix-os>` is any of
    `["freebsd", "linux", "netbsd", "redox", "illumos", "fuchsia", "emscripten", "android", "ios", "macos", "solaris"]`.
    This list needs to be updated if a new `unix` OS is supported by `rustc`'s official target list.
This would make the first example compile.

_Note:_ The contrapositive of these relations is also true.

Also, `target_family` is currently defined as not having mutually exclusive elements. This is because `target_family = "wasm"`
is not mutually exclusive with other target families. But, `target_family = "unix"` could be defined as mutually exclusive
with `target_family = "windows"` to increase usability. By extension, `target_family = "windows"` would now be mutually exclusive
with `target_os = "linux"`, for example.

_Note:_ More relations could be defined, for example `target_feature = "neon"` ⊆ `target_arch = "arm"`. With this however,
things start to get complicated.

## Misc

- Have `cargo add` check the `supported-targets` before adding a dependency.

> - Make this process part of the resolver.

Reviewer: What does this mean?  If its affecting `Cargo.lock`, then its affecting the resolver.

Author: Fixed in https://github.com/rust-lang/rfcs/pull/3759/commits/33c905ff3894a36fcbddc8c1a352ddefa1a16384.

- Show which targets are supported on `docs.rs`.
- Have search filters on `crates.io` for crates with support for specific targets.

> - Also add this field at the cargo-target level.

Reviewer: The fact that the alternatives call this out I feel like this needs to be under its own heading with a bit more explanation behind it

Author: Added in https://github.com/rust-lang/rfcs/pull/3759/commits/0cb167d80c60d2f6555b442a0a65a35895c5614c.

enhance usability, but this is left as a [future possibility](#more-cfg-relations).

## Behavior with unknown entries (and custom targets)

Just as `[target.my-custom-target.dependencies]` is allowed by `cargo`, `supported-targets` can contain
unknown entries. This is important because users may have different `rustc` versions, and the set of
official `rustc` targets is unstable; Targets can change name or be removed. Also, developers 
may want to support their custom target.

To determine if an entry in `supported-targets` is a target name or a `cfg` specification, the
same mechanism as for `[target.'cfg(..)']` is used (using
[`cargo-platform`](https://docs.rs/cargo-platform/latest/cargo_platform/index.html)).

## Ignoring builds for unsupported targets

When the target used is not supported by the package being built, the package will either be
skipped or an error will be raised, depending on how `cargo` was invoked. If cargo is invoked
in a workspace or virtual workspace without specifying a specific package, then `cargo` skips the package.
If a specific package is specified using `--package`, or if `cargo` is invoked on a single package,
then an error is raised.

> ## Eliminating unused dependencies from `Cargo.lock`

Reviewer: Should this be moved to a future possibility to better focus this RFC?

Author: I feel like this is an important motive of the RFC, even more so than raising errors on unsupported targets, because this can already be done as showed in prior art. I first wanted it to be a future possibility (as a two part implementation), but I got the feeling that people would prefer this as a lint, which is not compatible with dependency pruning.

Reviewer: A rule I've developed for PRs is to limit them to one controversial aspect.  @joshtriplett has extended this to RFCs which is an important insight.

By having a single RFC take on
- Build-target filtering
- Improving incompatible platform errors
- Pruning of `Cargo.lock` / `cargo vendor`

We have to solve every detail of each of those problems at once.  This usually manifests by one concern taking everyone's concern until its resolved and then we uncover the next layer of concerns and resolve one, etc until either reviewers are exhausted and miss a topic or the proposer is burnt out and gives up.

If we peel off just build-target filtering, we dramatically simplify this conversation which will make it much smoother to get this RFC approved and merged.  We can then use that as a base for iterating and experimenting on the other points.

Author: Oh, I thought you were referring to dependency pruning (the last line in the review window). I would be happy with moving build-target filtering to a future possibility.

Reviewer: I think my comment was misread.  I have highlighted `Cargo.lock` and said this should be deferred.  I am referring to deferring dependency pruning

- I suspect build-target filtering has the least controversy to it and could help us get the syntax established
- Incompatible platform errors have been controversial so far and I discuss splitting that out at https://github.com/rust-lang/rfcs/pull/3759#discussion_r1913515228
- `Cargo.lock` has a lot of design work with the Index (package and dependency target tracking) and a lot of implementation complexity (we don't even have this for artifact dep `target` field)

Author: Ok, I see the benefits of this.

I'll keep only package filtering, moving both dependency pruning and error reporting to a future possibility.

errors.

Cargo's documentation should give clear guidance for when to use this field, and should not suggest
using it by default. In particular, we should steer users to use this when they have good reason to
believe the crate will not compile or work as expected (e.g. because it uses target-specific APIs),
and not use it merely for "I haven't personally tested this on other targets".

Even then, it will happen that crates unnecessarily limit their dependents and users, because of 
over restrictive `supported-targets`. So, users must be able to disable the lint or error.
To alleviate this, a flag like `--ignore-supported-targets` could be added to `cargo` to ignore the `supported-targets` of a
package, and a field like
```toml
[dependencies]
overrestrictive-dep = { version = "0.1.0", ignore-supported-targets = true }
```
could be added to ignore the `supported-targets` of a specific dependency. This functionality
could otherwise be added to the [`[patch]` table](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-patch-section),
or be added as mutable metadata in package registries

> ([related discussion](https://blog.rust-lang.org/inside-rust/2024/03/26/this-development-cycle-in-cargo-1.78.html#why-is-this-yanked)).

Author: A question I had related to the interop of ignoring errors and `Cargo.lock` trimming:

A package may add restrictions to `supported-targets` even though the package builds fine without the restrictions _for the sake of `Cargo.lock` trimming_. Let's say then someone builds the package for a target that is not accepted by ignoring `supported-targets` somehow (say with `--ignore-supported-targets`). Should the package still trim out dependencies?

I would argue that yes, because I think its important for a package to generate the same lockfile without regards for which target was selected. Also, there was clear intent behind the `supported-targets` specification, even if it may not be obvious why. I would argue that a user should be able to bypass the `supported-targets` of a package, but should not be able to change the behavior of the package (bringing in extra deps) without making modifications to the code. Similarly, a library should be able to guarantee which dependencies it has no matter the target used. So `ignore-supported-targets` on a dependency should not change the dependency graph of the dependency.

Reviewer 13: But then what happens if `cargo build` is run on an arch that's been trimmed out? Does it force-ignore `supported-targets` and amend `Cargo.lock`? Does it result in the generation of some kind of `Cargo.lock.local`? Does it result in every build potentially causing an implicit `cargo update`?

Reviewer: I feel like this is covered previously (https://github.com/rust-lang/rfcs/pull/3759#discussion_r1909687096) and was part of what I was asking to have included in https://github.com/rust-lang/rfcs/pull/3759#discussion_r1920458612

Author: > But then what happens if `cargo build` is run on an arch that's been trimmed out? Does it force-ignore `supported-targets` and amend `Cargo.lock`? Does it result in the generation of some kind of `Cargo.lock.local`? Does it result in every build potentially causing an implicit `cargo update`?

It does none of that, it builds as if the `supported-targets` were respected and prunes the dependencies accordingly. If the build fails or does not behave normally at runtime because a dependency is missing then the user should realize that the `supported-targets` were there for a reason.

- Feature Name: `supported-targets`
- Start Date: 2025-01-08
- Pre-RFC: [Rust
  internals](https://internals.rust-lang.org/t/pre-rfc-allow-packages-to-specify-a-set-of-supported-targets/21979)
- RFC PR: [rust-lang/rfcs#3759](https://github.com/rust-lang/rfcs/pull/3759)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

The word _target_ is extensively used in this document. The
[glossary](https://doc.rust-lang.org/cargo/appendix/glossary.html#target) defines its many meanings.
Here, _target_ refers to the "Target Architecture" for which a package is built. Otherwise, the

> terms "cargo-target" and "target-triple" are used in accordance with their definitions in the

Reviewer: ```suggestion
terms "cargo-target" and "target-tuple" are used in accordance with their definitions in the
```

Reviewer 5: https://github.com/rust-lang/rust/pull/125579

Author: Every instance of `target-triple` changed to `target-tuple` in https://github.com/rust-lang/rfcs/pull/3759/commits/14f589666a5163330289cfc8c93afa3cb3b5b698.

When a `cargo` build command (e.g. `check`, `build`, `run`, `clippy`) is run, it checks that the
selected target satisfies the `supported-targets` of the package being built. If it does not, the
package is skipped or an error is raised, depending on how [`cargo` was invoked](ignoring-builds).
However, `supported-targets` is _only_ checked for commands that take a `--target` option and does
not affect other commands (e.g., `cargo fmt`).

As this field is limited to local development, `cargo package` / `cargo publish` will strip it from `Cargo.toml`.
Including the field in the `.crate` file is left as a [future possibility](#future-possibilities) for now.

This field is subject to [workspace inheritance](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-package-table).

## Ignoring builds for unsupported targets
[ignoring-builds]: #igonring-builds-for-unsupported-targets

If cargo is invoked in a workspace or virtual workspace without specifying a package as
build-target, then `cargo` skips any package that does not support the selected target. If a package
is specified using `--package` or if `cargo` is invoked on a single package, and the selected target

> does not satisfy the `supported-targets` of the package, then an error is raised.

Reviewer: We should make it explicit that our intent is to mirror the behavior of `required-features`

Author: Added a sentence about it in https://github.com/rust-lang/rfcs/pull/3759/commits/1cc79ef333a3616e12fb751517027b8fb918209d.

# Drawbacks
[drawbacks]: #drawbacks

> - Once implemented this does not yet solve the problem of dependency pruning in `Cargo.lock`, which
>   is a common use case for this feature.

Reviewer: To me, it is misleading to state this hear and not list it as a high likelihood future possibility.  In trying to come up with a rewording, I was leaning towards just removing this
```suggestion
```

- This is the first step towards a target aware `cargo`, which may increase `cargo`'s complexity,
  and bring more feature requests along these lines.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

## Format

The `cfg` string format was chosen because of its simplicity and expressiveness.
Other formats can be considered:

> Using a list of `cfg` strings, and also accepting explicit target-triples:

Reviewer: etc

exact, and the resolver can determine if a platform-specific dependency can be pruned from the
dependency tree more easily, hence why the original proposal chose this format. Being even simpler
to implement, this alternative may not be expressive enough for the common use case. Packages rarely
support specific target-triples, rather they support/require specific target attributes. What would
likely happen is that packages would copy and paste the target-triple list matching their
requirements from somewhere or someone else. Every time a new target with the same attribute is
added, the whole ecosystem would have to be updated.

# Prior art
[prior-art]: #prior-art

Users can already select which packages they want to select in a workspace with the flags
`--package` and `--exclude`.

The `per-package-target` nightly feature defines the `force-target` field, which is supposed to
force the package to build for a specific target-triple. This does not interact well when used in
dependencies, as one would expect a dependency to be built for the same target as the package.
`supported-targets` supersedes `force-target` because instead of enforcing a single target, it

> enforces a set of targets.

Reviewer: If we're talking about Cargo prior art, we should call out `required-features` for its design parallels

Author: Added in https://github.com/rust-lang/rfcs/pull/3759/commits/1f086faa72bac10d405e33249c1a5a908b6d9b03.

Published crates have mainly used their documentation to specify which targets they support, or they
would leave it up to the user to infer it. Some crates also made use of compile time errors to
ensure that `cfg` requirements are met, for example:
```rust
#[cfg(not(any(…)))]
compile_error!("unsupported target cfg");
```
[`getrandom`](https://github.com/rust-random/getrandom/blob/9fb4a9a2481018e4ab58d597ecd167a609033149/src/backends.rs#L156-L160)
is an example of a crate utilizing this method.

In other system level languages, vendoring dependencies is a common practice, and the user would be
responsible for ensuring that the dependencies are compatible with the target. For interpreted
languages, this is a non-issue because any platform being able to run the interpreter can run the
package.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> - Should we strip the `cfg` prefix from the field e.g., `supported-targets = 'target_os = "linux"'`?

Reviewer: I don't think we should.  This makes the form of the syntax explicit, follows the pattern of how `cfg` syntax is used everywhere else, and gives us room to evolve this field

[future-possibilities]: #future-possibilities

## Ensuring proper use of dependencies

Complicated errors caused by packages and dependencies that are incompatible with the selected
target can be avoided by using the information in the `supported-targets` field. For example, a
warning or an error could be raised if a package uses a dependency that does not accept the package's
`supported-targets`:
```toml
[package]
name = "bar"
supported-targets = 'cfg(target_os = "windows")'
```
```toml
[package]
name = "foo"
supported-targets = 'cfg(target_os = "linux")'

[dependencies]
bar = "0.1.0"
```
Here, a compilation error helps by showing which dependency is incompatible with the package's
`supported-targets`, rather than a cryptic error message about missing parts of `std`, or runtime
errors.

Cargo's documentation should give clear guidance for when to use this field, and should not suggest
using it by default. In particular, we should steer users to use this when they have good reason to
believe the crate will not compile or work as expected (e.g. because it uses target-specific APIs),
and not use it merely for "I haven't personally tested this on other targets".

> Even then, it will happen that crates unnecessarily limit their dependents and users, because of 
> over restrictive `supported-targets`. So, users must be able to remove the lint or error.
> To alleviate this, a flag like `--ignore-supported-targets` could be added to `cargo` to ignore the `supported-targets` of a
> package, and a field like
> ```toml
> [dependencies]
> overrestrictive-dep = { version = "0.1.0", ignore-supported-targets = true }
> ```
> could be added to ignore the `supported-targets` of a specific dependency. This functionality
> could otherwise be added to the [`[patch]` table](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-patch-section),
> or be added as mutable metadata in package registries
> ([related discussion](https://blog.rust-lang.org/inside-rust/2024/03/26/this-development-cycle-in-cargo-1.78.html#why-is-this-yanked)).
> 
> One thing to keep in mind is that disabling the `supported-targets` check for a package or a dependency
> removes the ability to prune the dependency tree of a package. That is because disabling the lint/error
> is equivalent to the package or the dependency supporting all targets. So, either the `--ignore-supported-targets`
> flag has the ability to change lockfile generation (this pattern is advised against), or dependencies are still
> pruned as if the package's `supported-targets` were respected, but the lint/error is ignored. The lockfile
> could also store the state which was used to generate it, but this currently is not implemented in `cargo`.

Reviewer: The way this is structured anchors the conversation with one solution, is a bit hard to follow, and leaves out some details

```suggestion
Even then, it will happen that crates unnecessarily limit their dependents and users because of 
overly restrictive `supported-targets`.
Some options for handling this include
- Doing nothing, encouraging people to upstream patches
- Encourage `[patch]`ing the dependency
  - Requires managing a fork
  - Every dependent of the package with a questionable `supported-targets` must do this
- Encourage unidiff `[patch]`es
  - Design has unresolved questions ([cargo#4648](https://github.com/rust-lang/cargo/issues/4648))
  - Every dependent of the package with a questionable `supported-targets` must do this
- A bespoke manifest override
  - One-off feature that needs design work
- A CLI override like `--ignore-rust-version`
  - This precludes `Cargo.lock` trimming as the lockfile is meant to capture dependencies for every potential state a package may be run in
  - This affects the entire dependency tree and not just the package with questionable `supported-targets`
  - Every dependent of the package with a questionable `supported-targets` must do this
- A lint like proposed for `package.rust-version`
  - Blocked on [cargo#12235](https://github.com/rust-lang/cargo/issues/12235)
  - See also CLI override
- Allow a registry database to override `supported-targets`
  - Blocked on a lot of design work ([related discussion](https://blog.rust-lang.org/inside-rust/2024/03/26/this-development-cycle-in-cargo-1.78.html#why-is-this-yanked))
```

force the package to build for a specific target-tuple. This does not interact well when used in
dependencies, as one would expect a dependency to be built for the same target as the package.
`supported-targets` supersedes `force-target` because instead of enforcing a single target, it
enforces a set of targets.

Published crates have mainly used their documentation to specify which targets they support, or they
would leave it up to the user to infer it. Some crates also made use of compile time errors to
ensure that `cfg` requirements are met, for example:
```rust
#[cfg(not(any(…)))]
compile_error!("unsupported target cfg");
```
[`getrandom`](https://github.com/rust-random/getrandom/blob/9fb4a9a2481018e4ab58d597ecd167a609033149/src/backends.rs#L156-L160)
is an example of a crate utilizing this method.

In other system level languages, vendoring dependencies is a common practice, and the user would be
responsible for ensuring that the dependencies are compatible with the target. For interpreted
languages, this is a non-issue because any platform being able to run the interpreter can run the

> package.

Reviewer 14: I would not agree with this statement about interpreted languages. Platform-specific packages are not unusual in Python, Javascript, etc.

I did some quick reading to see similar examples in other package managers. Here's a few, but I'm sure there are others.

Caveat: I do not use almost any of these tools, most of this is just based on reading.

## Python

Python has [platform compatibility tags](https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/). Not sure what impact those have.

## npm (Javascript)

Has [os](https://docs.npmjs.com/cli/v11/configuring-npm/package-json#os) and [cpu](https://docs.npmjs.com/cli/v11/configuring-npm/package-json#cpu) fields among other things. IIUC, these generate warnings if your platform does not match.

## Bazel

Has a [`target_compatible_with`](https://bazel.build/extending/platforms#skipping-incompatible-targets) field.

## Buck

Has a [`target_compatible_with`](https://buck2.build/docs/rule_authors/configurations/#target-platform-compatibility) field.

## Swift

[`Package.platforms`](https://developer.apple.com/documentation/packagedescription/package/platforms) can be used to specify supported platforms.

Reviewer: iiuc Python's compatibility tags is for matching a pre-built artifact to your system.  If one doesn't match, the source artifact is used.  This only works as a way to declare platform compatibility if you don't provide source.

Instead, people specify [trove classifiers](https://pypi.org/classifiers/).  I've not dug too deeply to see if they are informational only or if they have a behavioral impact.

Author: You are right, I assume it was the case without doing much research. I have added a section that discusses this in _Prior Art_, using your examples.

Author: I'll resolve this conversation as that was added in the latest commits.

version = "0.1.0"
edition = "2021"
supported-targets = 'cfg(any(target_os = "linux", target_os = "macos"))'
```
Here, only targets with the `linux` OS or the `macos` OS, are allowed to build the package. User
experience is enhanced by raising an error that fails compilation when the supported targets of a
package are not satisfied by the selected target.

This feature should be used when a package clearly does not support all targets. For example:
`io-uring` requires `cfg(target_os = "linux")`, `gloo` requires `cfg(target_family = "wasm")`, and
`riscv` requires `cfg(any(target_arch = "riscv32", target_arch = "riscv64"))`.

This feature increases cargo's knowledge of a package. For example, when working in a workspace
where some packages are for a platform with `target_os = "none"`, and some others are tools that
require a desktop OS, using `supported-targets` makes `cargo <command>` ignore packages which have
`supported-targets` that are not satisfied by the selected target.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

The `supported-targets` field is an optional key that tells cargo which targets the package can be

> built for.

Reviewer: What about build scripts?  Host targets can be very different than deployment targets.

Or when we do dependency filtering, do we only filter normal/dev dependencies and not build scripts?

Author: In theory any host-tuple should be able to run the build script, so build dependencies should support all targets. Enforcing this is very impractical, so we resort to failing the build if one of the `[build-dependencies]` does not support the host. See [here](https://github.com/rust-lang/rfcs/pull/3759/files#diff-7ca74978b782c1ea19b9cafec0cd71f3936a4791ac30656f361896ee217d391cR343-R352).

I also added some details in https://github.com/rust-lang/rfcs/pull/3759/commits/bc7d86bd8665898ce1fda3415308fe9aa04420c7, specifically that we do not filter build-dependencies.

The `per-package-target` nightly feature defines the `force-target` field, which is supposed to
force the package to build for a specific target-tuple. This does not interact well when used in
dependencies, as one would expect a dependency to be built for the same target as the package.
`supported-targets` supersedes `force-target` because instead of enforcing a single target, it
enforces a set of targets.

Published crates have mainly used their documentation to specify which targets they support, or they
would leave it up to the user to infer it. Some crates also made use of compile time errors to
ensure that `cfg` requirements are met, for example:
```rust
#[cfg(not(any(…)))]
compile_error!("unsupported target cfg");
```
[`getrandom`](https://github.com/rust-random/getrandom/blob/9fb4a9a2481018e4ab58d597ecd167a609033149/src/backends.rs#L156-L160)
is an example of a crate utilizing this method.

In other system level languages, vendoring dependencies is a common practice, and the user would be
responsible for ensuring that the dependencies are compatible with the target.

Some higher-level languages and build tools have the ability to specify which platforms are compatible.

> - Python has [platform compatibility tags](https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/#platform-compatibility-tags).
>     The reference explains how these are [used](https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/#use)
>     by installers to determine which build of a package to install. PyPI also has [classifiers](https://pypi.org/classifiers/)
>     which allows to add metadata (including which platforms are allowed) to a package.

Reviewer: Generally with bullets like this, people don't expect new information to be added and can miss classifiers being mentioned in here.  Better to have it as its own item
```suggestion
- Python package has [classifiers](https://pypi.org/classifiers/) as package metadata that includes supported platforms
- Python wheels (pre-built packages) have [platform compatibility tags](https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/#platform-compatibility-tags).
    The reference explains how these are [used](https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/#use)
    by installers to determine which build of a package to install.
```

- `npm` allows specifying which [`os`](https://docs.npmjs.com/cli/v11/configuring-npm/package-json#os) and
    [`cpu`](https://docs.npmjs.com/cli/v11/configuring-npm/package-json#cpu) a package supports. These generate an
    error when installing a package that does not support the platform used.
- Swift has [`package.platforms`](https://developer.apple.com/documentation/packagedescription/package/platforms), which
    allows specifying which platforms and versions a package support (mostly for apple products e.g., `macOS`, `iOS`, `watchOS`, `tvOS`).
- [Buck](https://buck2.build/docs/rule_authors/configurations/#target-platform-compatibility)
    and [Bazel](https://bazel.build/reference/be/common-definitions#common.target_compatible_with)
    both provide `target_compatible_with`.
Some accept a string or list of strings representing the platforms, while `Buck` & `Bazel` seem to accept a
form comparable to `cfg` in Rust.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- Should we strip the `cfg` prefix from the field e.g., `supported-targets = 'target_os = "linux"'`?

# Future possibilities
[future-possibilities]: #future-possibilities

> ## Ensuring proper use of dependencies

Reviewer: Strictly speaking, this feature is not blocked on this RFC.  We could do set logic between `target.*.dependencies` and `target.*.dependencies`.  When a package itself is only supported on some platforms, they could workaround the lack of `supported-targets` by putting *all* dependencies in a `target.*.dependencies`.

declared under. For example:
```toml
[package]
# ...
supported-targets = 'cfg(target_os = "linux")'

[target.'cfg(target_pointer_width = "64")'.dependencies]
foo = "0.1.0"
```
Here, it would suffice for `foo` to support `cfg(all(target_os = "linux", target_pointer_width =
"64"))`.

This would ensure that a package properly uses dependencies that are not available on all targets.
Assuming that the crate `io-uring` has `supported-targets = 'cfg(target_os = "linux")'`, a crate
could depend on it using:
```toml
[package]
# ...

[target.'cfg(target_os = "linux")'.dependencies]
io-uring = "0.1.0"
```
This would not be required if the package itself had `supported-targets = 'cfg(target_os =
"linux")'`, or an even stricter set.

### Artifact dependencies

If an artifact dependency has a `target` field, then the dependency would not be checked against the
package's `supported-targets`. However, the selected `target` for the dependency would need to be
compatible with the dependency's `supported-targets`, or else an error is raised. If the artifact
dependency does not have a `target` field, then it would be checked against the package's
`supported-targets`, like any other dependency.

> ## Eliminating unused dependencies from `Cargo.lock`

Reviewer: The main focus of this is limiting inclusion in `cargo vendor` so that should be clearly called out here

Reviewer 15: Ref: https://github.com/rust-lang/cargo/issues/7058

> ## Eliminating unused dependencies from `Cargo.lock`

Reviewer: Strictly speaking, this feature is not blocked on this RFC.  We could do set logic between `target.*.dependencies` and `target.*.dependencies`.  When a package itself is only supported on some platforms, they could workaround the lack of `supported-targets` by putting *all* dependencies in a `target.*.dependencies`.

Author: > When a package itself is only supported on some platforms, they could workaround the lack of supported-targets by putting all dependencies in a target.*.dependencies.

Indeed. `supported-targets` only adds an implied `target.*` bound.
Similarly people can already restrain their crate to specific targets with `compile_error!()`.

In the end `supported-targets` is only about the ergonomics of doing so and running commands like `cargo check --workspace`.

Reviewer: Could we add this to alternatives that the resolver changes could be made without manifest changes, just with less ergonomics?

> ## Eliminating unused dependencies from `Cargo.lock`

Reviewer: I wonder if we should not have `cargo vendor` filtering use `supported-targets` but instead a new `resolver.targets` `.cargo/config.toml` field.

Most of the rest of the RFC works equally for application and library developers.  This section is almost exclusively for application developers as they are the ones most likely to use `cargo vendor`.  These vendored dependencies are meant for upstream pre-built binaries.  They will be dealing with specific `--target <platform>`.  Any downstream-built binaries that might be beyond this set will likely also exclude the vendored dependencies.

So if we provide a different mechanism for this feature focused on target tuples
- Application developers are already coupled to target tuple specifics so this does not make their use of volatile target tuples any worse
- By being able to use target tuples, this removes the largest piece of volatility / complication
- Being further out of the way and in an already a volatile situation (target tuples) might mean its ok for any remaining volatility in the set relation logic?
- However, `Cargo.lock` would become dependent on transient / environment configuration.  We may want to record the targets used for resolving in `Cargo.lock` so they show up in any diffs
- However, `cargo publish` would include a `Cargo.lock` that is not intended for all platforms and `cargo install --locked` may fail.  Maybe `cargo publish` could resolve the published `Cargo.lock` for all platforms (unless `--locked` is specified for which it would error) and it would be ok?

So the question then is if this RFC is justifiable enough on its own without the expectation of this future possibility building on it.   I *think* so because the workarounds for `cargo check --workspace` with platform-specific packages is quite annoying, it could be useful for documentation purposes, and the future possibility of the lints could help people better manage their dependencies.

Reviewer 16: > However, Cargo.lock would become dependent on transient / environment configuration.
>
> a Cargo.lock that is not intended for all platforms

Please learn from the disaster that `npm` went through and do not repeat their mistake: https://github.com/npm/cli/issues/4828

Reviewer: It sounds like that was a default behavior while this is an opt-in buried in the config.  We'd be particularly intending this for applications that only support specific platforms.

A package's dependencies may themselves have `[target.'cfg(..)'.dependencies]` tables, which may
never be used because of the `supported-targets` restrictions of the package. These can safely be
eliminated from the dependency tree of the package.

Consider the following example:
```toml
[package]
name = "foo"
# ...
supported-targets = 'cfg(target_os = "linux")'

[dependencies]
bar = "0.1.0"
```
```toml
[package]
name = "bar"

[target.'cfg(target_os = "macos")'.dependencies]
baz = "0.1.0"
```
Currently, `baz` is included in the dependency tree of `foo`, even though `foo` is never built for
`macos`. `baz` could be pruned from the dependency tree of `foo`, since `target_os = "macos"` is
mutually exclusive with `target_os = "linux"`.

Formally, dependencies (and transitive dependencies) under `[target.**.dependencies]` tables are
eliminated from the dependency tree of a package if the `supported-targets` of the package is

> mutually exclusive with the target preconditions of the dependency.

Reviewer: The way this is written, it sounds like all `target.*.dependencies` tables are compared only to `supported-targets`

You could have a situation like
```toml
[package]
name = "foo"
supported-targets = 'cfg(target_os = "linux")'

[dependencies]
bar = "1.0.0"
```
```toml
[package]
name = "bar"

[target.'cfg(target_os="windows")'.dependencies]
baz = "1.0.0"
```
```toml
[package]
name = "baz"

[target.'cfg(target_os="linux")'.dependencies]
bob = "1.0.0"
```
```toml
[package]
name = "bob"
```

As specified, `bob` will be included in `Cargo.lock` / `cargo vendor` but it doesn't need to be.

However, implementing this properly in the resolver is complex and we may not want to do it for an MVP solution.

Author: If `baz` is filtered out then `bob` is also filtered out because its parent is removed from the tree. Maybe this need clarification.

Reviewer: The potential for both approaches should be called out so we don't forget when we get to this and because part of this is being evaluated based on these future possibilities.

### Comparing `supported-targets`

To prune the dependency tree, and to ensure proper use of dependencies, it becomes necessary to
compare `supported-targets`. When comparing two sets of `supported-targets`, it is necessary to
know if one is a _subset_ of the other, or if both are _mutually exclusive_. To proceed, both
are flattened to the same representation, and they are then compared. This process is done

> internally, and does not affect the `Cargo.toml` file.

Reviewer: Technically we'll only compare `supported-targets` to `target.*.dependencies` for `Cargo.lock` / `cargo vendor`.

Author: We could also want to lint/warn for dependencies that do not support all of `supported-targets`. In which case we need to compare the `supported-targets` of the dependencies with the `supported-targets` of the package.

there is an `any` inside an `all`, the statement is split into multiple `all` statements. For
example,
```toml
supported-targets = 'cfg(all(target_os = "linux", any(target_arch = "x86_64", target_arch = "arm"))'
```
is transformed into
```toml
supported-targets = 'cfg(any(all(target_os = "linux", target_arch = "x86_64"), all(target_os = "linux", target_arch = "arm")))'
```
If an `all` contains an `all`, the inner `all` is flattened into the outer `all`.

The result of these transformations on a `cfg` specification is a union of `cfg` specifications that
either contains a single specification, or an `all` operator with no nested operators.

#### The subset relation

To determine if the `supported-targets` set "A" is a subset of another such set "B", the standard
mathematical definition of subset is used. That is, "A" is a subset of "B" if and only if each
element of "A" is contained in "B".

So each element of the union forming "A" is compared against each element of the union forming "B".
A `cfg(all(A, B, ...))` is a subset of a `cfg(all(C, D ...))`, if the list `C, D, ...` is a subset
of the list `A, B, ...`.

> _Note_: `cfg(A) == cfg(all(A))`.

Reviewer: Do `not`s need to be called out in relation to subset handling?  In some cases, it seems like they'll need to evaluate to `true` and that should probably be called out

Author: I don't understand the claim here. Is it that if A and B are mutually exclusive, then `B ⊆ not(A)`?

  `["freebsd", "linux", "netbsd", "redox", "illumos", "fuchsia", "emscripten", "android", "ios",
  "macos", "solaris"]`. This list needs to be updated if a new `unix` OS is supported by `rustc`'s
  official target list. This would make the first example compile.

_Note:_ The contrapositive of these relations is also true.

Also, `target_family` is currently defined as not having mutually exclusive elements. This is
because `target_family = "wasm"` is not mutually exclusive with other target families. But,
`target_family = "unix"` could be defined as mutually exclusive with `target_family = "windows"` to
increase usability. By extension, `target_family = "windows"` would now be mutually exclusive with
`target_os = "linux"`, for example.

_Note:_ More relations could be defined, for example `target_feature = "neon"` ⊆ `target_arch =
"arm"`. With this however, things start to get complicated.

## Lint against unused target-specific tables

If a package has:
```toml
[package]
name = "example"
# ...
supported-targets = 'cfg(target_os = "linux")'

[target.'cfg(target_os = "windows")'.dependencies]
# ...
```

> A lint could be added to highlight the fact that the `[target]` table is unused.

Reviewer: `cfg(false)` / `cfg(any())` are intentional and should be called out as an exception

Author: Added in https://github.com/rust-lang/rfcs/pull/3759/commits/8145112522f9f6a858c39564e65bc3872922837d.

- Feature Name: `supported-targets`
- Start Date: 2025-01-08
- Pre-RFC: [Rust
  internals](https://internals.rust-lang.org/t/pre-rfc-allow-packages-to-specify-a-set-of-supported-targets/21979)
- RFC PR: [rust-lang/rfcs#3759](https://github.com/rust-lang/rfcs/pull/3759)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

The word _target_ is extensively used in this document. The
[glossary](https://doc.rust-lang.org/cargo/appendix/glossary.html#target) defines its many meanings.
Here, _target_ refers to the "Target Architecture" for which a package is built. Otherwise, the
terms "cargo-target" and "target-tuple" are used in accordance with their definitions in the
glossary.

> # Summary
> 
> The addition of `supported-targets` to `Cargo.toml`. This field is a `cfg` string that restricts the
> set of targets which a package supports. Packages can only be built for targets that satisfy their
> `supported-targets`.
> 
> ```toml
> [package]
> name = "hello_cargo"
> supported-targets = 'cfg(any(target_os = "linux", target_os = "macos"))'
> ```

Reviewer: Could you update the summary to clarify this is for filtering dependencies during local development?

Maybe something like:
````suggestion
# Summary

Workspace members will be able to specify their `supported-targets` in `Cargo.toml`, using `cfg` syntax, to avoid building those packages when using `--workspace`, much like `required-features` avoids building build-targets when the feature is not activated.

```toml
[package]
name = "hello_cargo"
supported-targets = 'cfg(any(target_os = "linux", target_os = "macos"))'
```
```console
$ cargo check --workspace --target x86_64-pc-windows-msvc
```
````
